src/Http/Http/src/StreamPipeReader.cs

```diff
@@ -0,0 +1,370 @@
+        private ExceptionDispatchInfo _exceptionInfo;
+        private object lockObject = new object();
+
+        private BufferSegment _readHead;
+        private int _readIndex;
+
+        private BufferSegment _commitHead;
+        private long _consumedLength;
+        private bool _examinedEverything;
+
+        private CancellationTokenSource InternalTokenSource
+        {
+            get
+            {
+                lock (lockObject)
+                {
+                    if (_internalTokenSource == null)
+                    {
+                        _internalTokenSource = new CancellationTokenSource();
+                    }
+                    return _internalTokenSource;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Creates a new StreamPipeReader.
+        /// </summary>
+        /// <param name="readingStream">The stream to read from.</param>
+        public StreamPipeReader(Stream readingStream) : this(readingStream, minimumSegmentSize: 4096)
+        {
+        }
+
+        /// <summary>
+        /// Creates a new StreamPipeReader.
+        /// </summary>
+        /// <param name="readingStream">The stream to read from.</param>
+        /// <param name="minimumSegmentSize">The minimum segment size to return from ReadAsync.</param>
+        /// <param name="pool"></param>
+        public StreamPipeReader(Stream readingStream, int minimumSegmentSize, MemoryPool<byte> pool = null)
+        {
+            _minimumSegmentSize = minimumSegmentSize;
+            _readingStream = readingStream;
+            _pool = pool ?? MemoryPool<byte>.Shared;
+        }
+
+        /// <inheritdoc />
+        public override void AdvanceTo(SequencePosition consumed)
+        {
+            AdvanceTo(consumed, consumed);
+        }
+
+        /// <inheritdoc />
+        public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
+        {
+            AdvanceTo((BufferSegment)consumed.GetObject(), consumed.GetInteger(), (BufferSegment)examined.GetObject(), examined.GetInteger());
+        }
+
+        private void AdvanceTo(BufferSegment consumedSegment, int consumedIndex, BufferSegment examinedSegment, int examinedIndex)
+        {
+            if (_isCompleted)
+            {
+                throw new InvalidOperationException("Reading is not allowed after reader was completed.");
+            }
+
+            if (consumedSegment == null)
+            {
+                return;
+            }
+
+            if (_readHead == null || _commitHead == null)
+            {
+                throw new InvalidOperationException("Pipe is already advanced past provided cursor.");
+            }
+
+            var returnStart = _readHead;
+            var returnEnd = consumedSegment;
+                
+            var consumedBytes = new ReadOnlySequence<byte>(returnStart, _readIndex, consumedSegment, consumedIndex).Length;
+            if (_consumedLength - consumedBytes < 0)
```

Comment:
Do you have a test for this condition? I think `Lenght` won't work if segments are not connected.

Reply:
When would segments not be connected?